system-tests-new/incremental/test_incremental.py

```diff
@@ -224,3 +228,158 @@ def assert_one_artefact(self, pb_keys, prebuild_groups, prebuild_folder,
+            ('a.456.foo', datetime(2022, 10, 1)),
+        ]
+
+        for a, t in artefacts:
+            path = folder / a
+            path.touch(exist_ok=False)
+            os.utime(path, (t.timestamp(), t.timestamp()))
+
+    # prune artefacts by age
+
+    def test_prune_oldest_15(self, tmp_path):
+        config, remaining = self._prune_oldest(tmp_path, age=15)
+        assert sorted(remaining) == [
+            config.prebuild_folder / 'a.123.foo',
+            config.prebuild_folder / 'a.234.foo',
+        ]
+
+    def test_prune_oldest_25(self, tmp_path):
+        config, remaining = self._prune_oldest(tmp_path, age=25)
+        assert sorted(remaining) == [
+            config.prebuild_folder / 'a.123.foo',
+            config.prebuild_folder / 'a.234.foo',
+            config.prebuild_folder / 'a.345.foo',
+        ]
+
+    # prune individual artefact versions (hashes) by age
+
+    def test_prune_versions_2(self, tmp_path):
+        config, remaining = self._prune_versions(tmp_path, n_versions=2)
+        assert sorted(remaining) == [
+            config.prebuild_folder / 'a.123.foo',
+            config.prebuild_folder / 'a.234.foo',
+        ]
+
+    def test_prune_versions_3(self, tmp_path):
+        config, remaining = self._prune_versions(tmp_path, n_versions=3)
+        assert sorted(remaining) == [
+            config.prebuild_folder / 'a.123.foo',
+            config.prebuild_folder / 'a.234.foo',
+            config.prebuild_folder / 'a.345.foo',
+        ]
+
+    # pruning everything not current
+
+    def test_prune_unused(self, tmp_path):
+        config, remaining = self._prune_unused(tmp_path)
+        assert sorted(remaining) == [
+            config.prebuild_folder / 'a.123.foo',
+            config.prebuild_folder / 'a.456.foo',
+        ]
+
+    # pruning a file which is covered by both the age and the version pruning code.
+    # this is to protect against trying to delete a non-existent file.
+    def test_prune_overlap(self, tmp_path):
+        config = BuildConfig(
+            project_label=PROJECT_LABEL, fab_workspace=tmp_path, multiprocessing=False,
+            steps=[
+                CleanupPrebuilds(older_than=timedelta(days=15), n_versions=2)
+            ],
+        )
+
+        # create artefacts
+        self.create_artefacts(config.prebuild_folder)
+
+        # run the cleanup
+        config.run()
+
+        # return the remaining files
+        remaining_artefacts = file_walk(config.prebuild_folder)
+        assert sorted(remaining_artefacts) == [
+            config.prebuild_folder / 'a.123.foo',
+            config.prebuild_folder / 'a.234.foo',
+        ]
+
+    # helpers
+
+    def _prune_oldest(self, tmp_path, age):
+        # common code to test pruning by age
+        config = BuildConfig(
+            project_label=PROJECT_LABEL, fab_workspace=tmp_path, multiprocessing=False,
+            steps=[
+                CleanupPrebuilds(older_than=timedelta(days=age))
+            ],
```

Comment:
You might be able to use a [parameterised test](https://docs.pytest.org/en/7.2.x/how-to/parametrize.html) to save some boilerplate. There may be a problem around interacting with the `test_Tree` fixture mentioned above.

Reply:
I did a bit of refactoring and was able to parameterise all but one into a single test. It's much more concise, and reads better too now, I think. (Note: I've found it can be awkward to investigate and debug failing parameterised tests, especially when using breakpoints.)